src/nvim/os/shell.c

```diff
@@ -309,15 +318,134 @@ static void system_data_cb(Stream *stream, RBuffer *buf, size_t count,
+  static size_t     visit       = 0;  // "Pulse" count of the current throttle.
+  static size_t     max_visits  = 0;
+  static char       pulse_msg[] = { ' ', ' ', ' ', '\0' };
+
+  if (!size) {
+    // bool previous_decision = (visit > 0);
+    started = received = visit = 0;
+    max_visits = 20;
+    return true;
+  }
+
+  received += size;
+  if (received < OUT_DATA_THRESHOLD
+      // Display at least the first chunk of output even if it is big.
+      || (!started && received < size + 1000)) {
+    return false;
+  } else if (!visit) {
+    started = os_hrtime();
+  } else if (visit >= max_visits) {
+    uint64_t since = os_hrtime() - started;
+    if (since < NS_1_SECOND) {
+      // Adjust max_visits based on the current relative performance.
+      // Each "pulse" period should last >=1 second so that it is perceptible.
+      max_visits = (2 * max_visits);
+    } else {
+      received = visit = 0;
+      return false;
+    }
+  }
+
+  visit++;
+  // Pulse "..." at the bottom of the screen.
+  size_t tick = (visit == max_visits)
+                ? 3  // Force all dots "..." on last visit.
+                : (visit + 1) % 4;
+  pulse_msg[0] = (tick == 0) ? ' ' : '.';
+  pulse_msg[1] = (tick == 0 || 1 == tick) ? ' ' : '.';
+  pulse_msg[2] = (tick == 0 || 1 == tick || 2 == tick) ? ' ' : '.';
+  if (visit == 1) {
+    screen_del_lines(0, 0, 1, (int)Rows, NULL);
+  }
+  int lastrow = (int)Rows - 1;
+  screen_puts_len((char_u *)pulse_msg, ARRAY_SIZE(pulse_msg), lastrow, 0, 0);
+  ui_flush();
+  return true;
+}
+
+/// Saves or prints a chunk of throttled (skipped) output. Used to ensure the
+/// last ~page of output for a shell-command is always displayed.
+///
+/// Init mode: Resets the internal state.
+///   output = NULL
+///   size   = 0
+/// Print mode: Displays the previous saved data.
+///   output = NULL
+///   size   = SIZE_MAX
+///
+/// @param  output  Data to save, or NULL to invoke a special mode.
+/// @param  size    Length of `output`.
+static void out_data_throttled_save_or_print(char *output, size_t size)
+{
+#define MAX_CHUNK_SIZE (OUT_DATA_THRESHOLD / 2)
+  static char    last_skipped[MAX_CHUNK_SIZE];  // Saved output.
+  static size_t  last_skipped_len = 0;
+
+  assert(output != NULL || (size == 0 || size == SIZE_MAX));
+
+  if (output == NULL && size == 0) {          // Init mode
+    last_skipped_len = 0;
+    return;
+  }
+
+  if (output == NULL && size == SIZE_MAX) {   // Print mode
+    out_data_append_to_screen(last_skipped, last_skipped_len, false);
+    return;
+  }
+
+  if (size >= MAX_CHUNK_SIZE) {               // Save mode
+    size_t start = size - MAX_CHUNK_SIZE;
+    memcpy(last_skipped, output + start, MAX_CHUNK_SIZE);
+    last_skipped_len = MAX_CHUNK_SIZE;
+  } else if (size > 0) {
+    // Length of the old data that can be kept.
+    size_t old_keep_len = MAX_CHUNK_SIZE - size;
+    size_t old_keep_start = last_skipped_len - old_keep_len;
```

Comment:
@oni-link Thanks for noticing that, see 07dc5592b2e0b952f9a8aa7d72001eefd073aeb9. This fixed behavior I noticed with certain output patterns.